woocommerce/woocommerce-ios
# Login: Install Woo from the store picker

Closes: #7429 


### Description


This PR improves the store picker by including non-woo sites in the list. Tapping on a non-woo site will navigate to the No Woo error screen, where WooCommerce can be installed.

Since this PR changes all logic regarding to the table view, I added a view model for the `StorePickerViewController` so that the logic can be tested easily.

Other changes: 
- `preselectStoreIfPossible` was updated to only preselect a store if it has WooCommerce.
- Added a new property `num_of_non_woo_sites` when tracking store picker view.
- Updated `StoreTableViewCell` to show warning icons for non-woo sites.
- The store list now sorts all stores on the top and non-Woo sites at the bottom. The list sections are named based on whether the sites have Woo.
- The only store picker that doesn't show non-Woo sites is the one for switching store when logged in.

Internal discussion: p1659946510389419-slack-C03L1NF1EA3

### Testing instructions

Prerequisite: make sure your WP.com account have access to at least one Woo Store and at least one non-Woo sites.
**Testing WP.com login flow**
1. Log out from your account if necessary, then select Continue with WP.com.
2. Log in with your WP.com credentials.
3. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. In Xcode console, notice both the number of stores and non-woo sites are tracked: `🔵 Tracked site_picker_stores_shown, properties: [AnyHashable("num_of_non_woo_sites"): 3, AnyHashable("num_of_stores"): 1]`
4. Select a non-woo site that is self-hosted or Atomic. Notice that a No Woo error screen is displayed, with the Install WooCommerce button present. The "See connected stores" button shouldn't be present here. Notice in Xcode console: `🔵 Tracked site_picker_non_woo_site_tapped, properties: [AnyHashable("is_non_atomic"): false]`
5. Select Install WooCommerce. When the installation completes, the No Woo screen is dismissed and the store list is updated with the updated store. This store should be selected automatically.
6. If you have a non-Atomic site, selecting it from the store picker should display the No Woo error screen without the Install WooCommerce button. Xcode console should log `🔵 Tracked site_picker_non_woo_site_tapped, properties: [AnyHashable("is_non_atomic"): true]`
7. If you have a JCP site, selecting it from the store picker should display the No Woo error screen without the Install WooCommerce button.

**Testing the Account Mismatch flow**
1. Log out from your account if necessary, then select Enter a site address.
2. Enter the address for a Woo store to which you don't have access.
3. Proceed to log in with your WP.com account.
4. Notice that the Account Mismatch screen is displayed. Select "See connected stores" button.
5. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the No Woo flow**
1. Log out from your account if necessary, then select Enter a site address.
2. Enter the address for a non-woo site that you have access to.
3. Proceed to log in with your WP.com account.
4. Notice that the Account Mismatch screen is displayed. Select "See connected stores" button.
5. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the app start flow**
1. Log out from your account if necessary, then select Continue with WP.com.
2. Log in with your WP.com credentials.
3. Notice that you are navigated to the store picker. Kill the app.
4. Open the app again, notice that a store picker is displayed with both Woo and non-Woo sites.
5.  Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the switching store flow**
1. Log in to your WP.com account and select a Woo store.
2. In the Settings screen, select the "Switch store" button.
3. Notice that the store picker displays only woo stores.

Please feel free to also test the case for a WP.com account without either Woo or non-woo sites. The empty screen should be displayed as before.


### Screenshots


https://user-images.githubusercontent.com/5533851/183809469-9f22d97b-6902-48f4-a607-577da717502d.mp4

Switching store:

<img src="https://user-images.githubusercontent.com/5533851/183815498-25109997-4352-479c-86a5-08efd34e72b2.png" width=320 />

No sites:

<img src="https://user-images.githubusercontent.com/5533851/183834565-dfaff29b-ed71-4545-a3ff-c3ef6cfae25e.png" width=320 />


---
